### DISCUSSION TITLE
Deploying modules in Azure via CI/CD pipeline.

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#1127] [Q&A] [answered]
We are setting up CI/CD pipelines in Azure deveops and have three repos:  One for the framework, one for modules, and one for the theme.  For the theme and modules repos, we'd like to build the Nuget packages and then deploy them into the framework.  This will allow is to automate the deployments and not have to manually upload modules.  Does anyone have an example of this, or knows where to put the packages?  I looked at the framework code and I think it's looking for package files in the bin folder.  There is a process that scans that folder.  I'm also wondering how that code gets kicked off.

**Comment by sbwalker** [comment, marked as answer]
@markdav-is if you place your module nupkg packages in the wwwroot/Modules folder they will be automatically installed during Oqtane startup. Similarly if you place your theme nupkg packages in the wwwroot/Themes folder they will be automatically installed. You will notice the release.cmd file generated by the Module Creator deploys the nupkg to this location as well:

XCOPY "*.nupkg" "..\..\oqtane.framework\Oqtane.Server\wwwroot\Modules\" /Y

**Reply by markdav-is** [reply]
Awesome thanks!

**Reply by markdav-is** [reply]
This is working pretty well.  I'm curious about the "installed during startup" and if there is a preferred way to restart a running Oqtane instance.  We can do this is a brute-force way in Azure.  Is there mechanism in Oqtane to kick off a restart?

**Reply by sbwalker** [reply]
@markdav-is if you login to the application and go to the Admin Dashboard and select the System Info option, there is a Restart Application button:

![image](https://user-images.githubusercontent.com/4840590/109700925-62d2ef80-7b60-11eb-98bd-5a91a93b2474.png)

**Reply by markdav-is** [reply]
very nice!

**Reply by markdav-is** [reply]
I don't see that button in ver 2.0.0 so we'll pull 2.0.1 next sprint!

**Reply by bfreed76** [reply]
Last night we deployed a modules nupkg package to our QA site last night and it killed the server so we can't access the Oqtane dashboard anymore. I can get into the server file system via the Kudu Console but this doesn't give me a wwwroot/modules path. Do you know of another way to access wwwroot/modules via Azure?

![Screenshot 2022-12-29 091842](https://user-images.githubusercontent.com/77690150/209987221-e66cfcaa-979f-4f3a-b263-4174757ea327.png)

**Reply by bfreed76** [reply]
We figured it out! I connected WinSCP using FTPS credentials info in the App Service > Deployment Center and dropped the fixed nupkg in /site/wwwroot/wwwroot/Modules and reset through Azure. Thanks!

**Comment by reteid** [comment]
Hi @markdav-is , 

Im also setting up a build pipeline for my Oqtane Modules in Azure DevOps. 

Seems like the build steps are something like : 

- Download code
- restore Nugets
- Build indivual CS Projects (cant build the Solution as the Oqtane Framework reference will be missing)
- Copy from $(agent.builddirectory)\Package\**.nupkg      to      $(build.artifactstagingdirectory)

Have you perhaps got a YAML file of the steps ?
thanks in anticipation

**Reply by ghost** [reply]
@reteid Did you figure this out? I'm also thinking about setting up a pipeline. Can you share your yaml file?

**Reply by markdav-is** [reply]
We have not set this up in production, but hope to do this for a few projects soon.  I'll try and remember to share in this thread when I get it working.

**Reply by bfreed76** [reply]
@reteid did you ever get it to work?

**Reply by reteid** [reply]
No sorry I never got around to it.. please share if you manage to get it right
